pyhf/__init__.py

```diff
@@ -180,63 +153,46 @@ def __init__(self, spec, **config_kwargs):
-
-    def _histosysdelta(self, channel, sample, pars):
-        modifiers = [m['name'] for m in sample['modifiers']
-                if m['type'] == 'histosys']
-        stack = None
-        for m in modifiers:
-            modifier, modpars = self.config.modifier(m), pars[self.config.par_slice(m)]
-            assert int(modpars.shape[0]) == 1
-
-            # print 'MODPARS', type(modpars.data)
-            mod_delta = _hfinterp_code0(modifier.at_minus_one[channel['name']][sample['name']],
-                                        modifier.at_zero[channel['name']][sample['name']],
-                                        modifier.at_plus_one[channel['name']][sample['name']],
-                                        modpars)[0]
-            stack = tensorlib.stack([mod_delta]) if stack is None else tensorlib.stack([stack,mod_delta])
-
-        return tensorlib.sum(stack, axis=0) if stack is not None else None
-
     def expected_sample(self, channel, sample, pars):
         # for each sample the expected ocunts are
         # counts = (multiplicative factors) * (normsys multiplier) * (histsys delta + nominal hist)
         #        = f1*f2*f3*f4* nomsysfactor(nom_sys_alphas) * hist(hist_addition(histosys_alphas) + nomdata)
         # nomsysfactor(nom_sys_alphas)   = 1 + sum(interp(1, anchors[i][0], anchors[i][0], val=alpha)  for i in range(nom_sys_alphas))
         # hist_addition(histosys_alphas) = sum(interp(nombin, anchors[i][0],
         # anchors[i][0], val=alpha) for i in range(histosys_alphas))
-        nom = tensorlib.astensor(sample['data'])
-        histosys_delta = self._histosysdelta(channel, sample, pars)
-
-        interp_histo = tensorlib.sum(tensorlib.stack([nom, histosys_delta]), axis=0) if (histosys_delta is not None) else nom
-
+        #
+        # Formula:
+        #     \nu_{cb} (\phi_p, \alpha_p, \gamma_p) = \lambda_{cs} \gamma_{cb} \phi_{cs}(\alpha) \eta_{cs}(\alpha) \sigma_{csb}(\alpha)
+        # \gamma == statsys, shapefactor
+        # \phi == normfactor, overallsys
+        # \sigma == histosysdelta + nominal
+
+        # first, collect the factors from all modifiers
+        results = {'shapesys': [],
+                   'normfactor': [],
+                   'shapefactor': [],
+                   'staterror': [],
+                   'histosys': [],
+                   'normsys': []}
+        for m in sample['modifiers']:
+            modifier, modpars = self.config.modifier(m['name']), pars[self.config.par_slice(m['name'])]
+            results[m['type']].append(modifier.apply(channel, sample, modpars))
+
+
+        # start building the entire set of factors
         factors = []
-        factors += self._multiplicative_factors(channel, sample, pars)
-
-        # print('sample factors {}'.format(factors))
-
+        factors += results['shapesys']
+        factors += results['normfactor']
+        factors += results['shapefactor']
+        factors += results['staterror']
+        factors += results['normsys']
+
```

Comment:
there are two times of factors: vectors and scalars that get broadcasted, perhaps it's useful in the code to have comments which are which

`shapesys`, `shapefactor`, staterror` are vectors

`normfactor` , `normsys` ae scalars

Reply:
Can you explain this more? I don't get the broadcasting, and I don't see how `[a, [b,c], d]` works correctly? Is broadcasting sort of trying to flatten things out or do something else more magically?

Reply:
I think we lost the documentation on this when we introduce the tensor backends. The idea is that we compute all bin-values at once.. each bin is a product of various factors, but sum are per-channel the other per-channel

```
b1 = shapesys_1   |      shapef_1   |
b2 = shapesys_2   |      shapef_2   |
...             normfac1    ..     normfac2
                (broad)            (broad)
bn = shapesys_n   |      shapef_1   |
```

this can be achieved by `numpy`'s `broadcast_arrays` and `np.product`

```
np.product(np.broadcast_arrays([2],[3,4,5],[6],[7,8,9]),axis=0)
```

is `[ 2*3*6*7, 2*4*6*8, 2*5*6*9]`